producer/types.go

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package producer
+
+import (
+	"github.com/m3db/m3cluster/services"
+)
+
+// DataFinalizeReason defines the reason why the data is being finalized by Producer.
+type DataFinalizeReason int
+
+const (
+	// Consumed means the data has been fully consumed.
+	Consumed DataFinalizeReason = iota
+
+	// Expired means the data has been expired.
+	Expired
+)
+
+// Data contains the data that will be produced by the producer.
+// It should only be finalized by the producer.
+type Data interface {
+	// Shard returns the shard of the data.
+	Shard() uint32
+
+	// Bytes returns the bytes of the data.
+	Bytes() []byte
+
+	// Size returns the size of the bytes of the data.
+	Size() uint32
+
+	// Finalize will be called by producer to indicate the end of its lifecycle.
+	Finalize(DataFinalizeReason)
+}
+
+// Producer produces data to a topic.
+type Producer interface {
+	// Produce produces the data.
+	Produce(data Data) error
+
+	// RegisterFilter registers a filter to a consumer service.
+	RegisterFilter(sid services.ServiceID, fn FilterFunc)
+
+	// RemoveFilter removes the filter of a consumer service.
+	RemoveFilter(sid services.ServiceID)
+
+	// Init initializes a producer.
+	Init()
+
+	// Close stops the producer from accepting new requests immediately.
+	// It will also block until all the data buffered in Producer has been consumed.
+	Close()
+}
+
+// FilterFunc can filter data.
+type FilterFunc func(data Data) bool
+
+// Options configs a producer.
+type Options interface {
+	// Buffer returns the buffer.
+	Buffer() Buffer
+
+	// SetBuffer sets the buffer.
+	SetBuffer(value Buffer) Options
+
+	// Writer returns the writer.
+	Writer() Writer
+
+	// SetWriter sets the writer.
+	SetWriter(value Writer) Options
+}
+
+// Buffer buffers all the data in the producer.
+type Buffer interface {
```

Comment:
Do we want to define the buffer interface is `buffer/types.go`?

Reply:
Potentially, that means I need to move the implementation of producer interface to another package though, otherwise there will be a dependency circle that producer.Producer depends on buffer.Buffer which further depends on producer.Data